### PR TITLE
Fix NPE during search with source filtering if the source is disabled.

### LINKED ISSUE #7758: Better handling when attempting to search with source filtering against docs with disabled source
Reproducible using Java API 1.3.x.  See repro at :  https://gist.github.com/ppf2/1a14a31b3add6c450c10

In short, if you have a doc type with source disabled and if you attempt to use the Java API to search for docs of this type specifying source filtering via SearchSourceBuilder, you will receive an unintuitive exception:

```java
Exception in thread "main" org.elasticsearch.action.search.SearchPhaseExecutionException: Failed to execute phase [query_fetch], all shards failed; shardFailures {[bqSWDvHIT7WCtEJKymhnHA][disable_source][0]: ElasticsearchIllegalArgumentException[No matching content type for null]}
```

### PR DESCRIPTION
**Pull request by qwerty4030**
Fix NPE during search with source filtering if the source is disabled.
Instead of throwing an NPE, search response hits with source filtering will not contain the source if it is disabled in the mapping.

Tests pass: `gradle test` `gradle core:integTest`

Closes #7758

### CODE REVIEW

**Comment by s1monw** [comment]
I mean the code LGTM but I wonder if we can start adding a unittest instead of an integration test for this?

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhaseTests.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+        @Override
+        public SearchLookup lookup() {
+            SearchLookup lookup = super.lookup();
+            lookup.source().setSource(source);
+            return lookup;
+        }
+    }
+
+    public void test() throws IOException {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhaseTests.java]: Maybe rename this `testBasicOperation` or something?

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhaseTests.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        XContentBuilder source = XContentFactory.jsonBuilder().startObject()
+            .field("field", "value")
+            .endObject();
+        FetchSubPhase.HitContext hitContext = hitExecute(source, true, null, null);
+        Map<String, Object> expected = new HashMap<>();
+        expected.put("field", "value");
+        assertThat(hitContext.hit().sourceAsMap(), equalTo(expected));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhaseTests.java]: Personally I'd write this as `assertEquals(singletonMap("field", "value"), hitContext.hit().sourceAsMap());`. If you like it better the way you have it that is fine too.

**Comment by qwerty4030** [comment]
@s1monw added unit tests. reverted integration test for disabled source.
The unit tests cover the disabled source scenario as well as the existing integration tests. Should we remove some of the other integration tests? Also I didn't get into randomized testing, complex doc sources, etc... I think these should be covered by [XContentMapValuesTests](https://github.com/elastic/elasticsearch/blob/master/core/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java) (tests the actual filter logic).

**Comment by nik9000** [comment]
Looks good to me. I left a few style comments if you are interested in them, otherwise I can test locally and merge.

**Comment by s1monw** [comment]
test this please

**Comment by nik9000** [comment]
Awesome. @s1monw's kicked off a CI build for this.

**Comment by nik9000** [comment]
@elasticmachine, test this please.

**Pull request merged**